cpu/sam0_common/periph/dma.c

```diff
@@ -0,0 +1,334 @@
+        mutex_init(&dma_ctx[i].sync_lock);
+    }
+    /* Enable all priorities with RR scheduling */
+    DMAC->CTRL.reg = DMAC_CTRL_LVLEN0 |
+                     DMAC_CTRL_LVLEN1 |
+                     DMAC_CTRL_LVLEN2 |
+                     DMAC_CTRL_LVLEN3;
+    DMAC->PRICTRL0.reg = DMAC_PRICTRL0_RRLVLEN0 |
+                         DMAC_PRICTRL0_RRLVLEN1 |
+                         DMAC_PRICTRL0_RRLVLEN2 |
+                         DMAC_PRICTRL0_RRLVLEN3;
+
+    DMAC->BASEADDR.reg = (uint32_t)descriptors;
+    DMAC->WRBADDR.reg = (uint32_t)writeback;
+
+#if defined(CPU_FAM_SAML11)
+    NVIC_EnableIRQ(DMAC_0_IRQn);
+    NVIC_EnableIRQ(DMAC_1_IRQn);
+    NVIC_EnableIRQ(DMAC_2_IRQn);
+    NVIC_EnableIRQ(DMAC_3_IRQn);
+#elif defined(CPU_FAM_SAMD5X)
+    NVIC_EnableIRQ(DMAC_0_IRQn);
+    NVIC_EnableIRQ(DMAC_1_IRQn);
+    NVIC_EnableIRQ(DMAC_2_IRQn);
+    NVIC_EnableIRQ(DMAC_3_IRQn);
+    NVIC_EnableIRQ(DMAC_4_IRQn);
+#else
+    NVIC_EnableIRQ(DMAC_IRQn);
+#endif
+
+    DMAC->CTRL.bit.DMAENABLE = 1;
+}
+
+dma_t dma_acquire_channel(void)
+{
+    dma_t channel = UINT8_MAX;
+    unsigned state = irq_disable();
+
+    if (channels_free) {
+        channel = bitarithm_lsb(channels_free);
+        /* Clear channel bit */
+        channels_free &= ~(1 << channel);
+        /* ensure the sync lock is locked */
+        mutex_trylock(&dma_ctx[channel].sync_lock);
+    }
+    irq_restore(state);
+    return channel;
+}
+
+void dma_release_channel(dma_t dma)
+{
+    unsigned state = irq_disable();
+#ifdef CPU_FAM_SAMD5X
+    DMAC->Channel[dma].CHCTRLA.reg = DMAC_CHCTRLA_SWRST;
+#else
+    DMAC->CHID.reg = dma;
+    /* Reset DMA channel */
+    DMAC->CHCTRLA.reg = DMAC_CHCTRLA_SWRST;
+#endif
+    channels_free |= 1 << dma;
+    irq_restore(state);
+}
+
+static inline void _set_source(DmacDescriptor *descr, void *src)
+{
+    descr->SRCADDR.reg = (uint32_t)src;
+}
+
+static inline void _set_destination(DmacDescriptor *descr, void *dst)
+{
+    descr->DSTADDR.reg = (uint32_t)dst;
+}
+
+static inline void _set_len(DmacDescriptor *descr, size_t len)
+{
+    descr->BTCNT.reg = len;
+}
+
+static inline void _set_next_descriptor(DmacDescriptor *descr, void *next)
+{
+    descr->DESCADDR.reg = (uint32_t)next;
+}
+
+static inline void _set_btctrl(DmacDescriptor *descr, uint8_t width)
+{
+    descr->BTCTRL.reg = width << DMAC_BTCTRL_BEATSIZE_Pos |
+                        DMAC_BTCTRL_VALID;
+}
+
+void dma_setup(dma_t dma, unsigned trigger, uint8_t prio, bool irq)
+{
+#ifdef CPU_FAM_SAMD5X
+    DMAC->Channel[dma].CHCTRLA.reg = DMAC_CHCTRLA_TRIGACT_BURST |
+                                     (trigger << DMAC_CHCTRLA_TRIGSRC_Pos);
+    DMAC->Channel[dma].CHPRILVL.reg = prio;
+    DMAC->Channel[dma].CHINTENCLR.reg = 0xFF;
+    if (irq) {
+        DMAC->Channel[dma].CHINTENSET.reg = DMAC_CHINTENSET_TCMPL;
+    }
+#else
+    /* Ensure that this set of register writes is atomic */
+    unsigned state = irq_disable();
```

Comment:
Yes, but the additional overhead of a mutex lock/unlock is imho not worth it here. With a mutex, the thread can possibly get interrupted in the middle of the modification, so there is two mutex locks, two unlocks and two context switches as overhead, where with the current construct it would be two irq_disable/restore pairs. Which is always going to be faster if you consider that both `mutex_unlock` and `mutex_lock` have a critical section inside them.